### PR TITLE
Enforcer rule to use new value class

### LINKED ISSUE #628: Enforcer rule to use new value class
Part of the refactoring (#574 )

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #628.

Updates to `toString` method helps the output of the enforcer rule message ([example in test case](https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/629/files#diff-86bd74d650f0820130645ec02069c5d6R319))

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassFile.java [hunk 1]
```diff
@@ -26,11 +26,11 @@
  * A tuple of a class name and a jar file to uniquely locate the class implementation in a class
  * path.
  */
-final class ClassFile {
+public final class ClassFile {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassFile.java]: Perhaps not in the cl, but maybe the right name for this is "CompiledClass" or something like that. In ther words this uniquely locates the byte code of a Java class using a class name and a jar path.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassFile.java [hunk 2]
```diff
@@ -26,11 +26,11 @@
  * A tuple of a class name and a jar file to uniquely locate the class implementation in a class
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassFile.java]: I don't think this is actually a tuple.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 3]
```diff
@@ -81,7 +85,7 @@ private LinkageChecker(
   }
 
   @VisibleForTesting
-  ImmutableSetMultimap<ClassFile, SymbolProblem> findSymbolProblems() {
+  public ImmutableSetMultimap<ClassFile, SymbolProblem> findSymbolProblems() {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: Still visible for testing?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java [hunk 4]
```diff
@@ -85,12 +84,25 @@ public int hashCode() {
...
-  public String toString() {
-    return MoreObjects.toStringHelper(this)
-        .omitNullValues()
-        .add("errorType", errorType)
-        .add("symbol", symbol)
-        .add("containingClass", containingClass)
-        .toString();
+  public final String toString() {
+    StringBuilder builder = new StringBuilder(symbol.toString());
+    switch (getErrorType()) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: Switches on type usually signal that the object hierarchy isn't being fully used. Here the relevant info can be pushed into ErrorType, and this becomes something like `getErrorType().getMessage()` with no switch.

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java [hunk 5]
```diff
@@ -123,34 +128,32 @@ public void execute(@Nonnull EnforcerRuleHelper helper) throws EnforcerRuleExcep
...
-                  .map(JarLinkageReport::reachableErrors)
-                  .collect(toImmutableList());
-          linkageReport = LinkageCheckReport.create(reachableErrorReports);
-          int reachableErrorCount =
-              reachableErrorReports.stream()
-                  .mapToInt(JarLinkageReport::getErrorCount)
-                  .sum();
-          errorCount = reachableErrorCount;
+          ClassReferenceGraph classReferenceGraph = linkageChecker.getClassReferenceGraph();
+          symbolProblems = symbolProblems.entries().stream().filter(
```
Comment (elharo) [inline enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java]: This is hard to follow. Can it be broken up into multiple smaller statements, possibly without reusing a variable?

Comment (suztomo) [reply enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java]: The indentation was incorrect. Does this look good now?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassFile.java (hunk 1 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassFile.java]: I feel `ClassFile` is still good because a class file in Java means it's compiled.

To clarify it, updated javadoc to `A compiled class file of {@code className} in {@code jar}...`.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassFile.java [hunk 6]
```diff
@@ -23,8 +23,8 @@
 import java.util.Objects;
 
 /**
- * A tuple of a class name and a jar file to uniquely locate the class implementation in a class
- * path.
+ * A compiled class file of {@code className} in {@code jar} to uniquely locate the class
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassFile.java]: Perahps, "a locator for a compiled class file in a jar"

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java [hunk 7]
```diff
@@ -160,7 +159,7 @@ public void execute(@Nonnull EnforcerRuleHelper helper) throws EnforcerRuleExcep
           }
         } else {
           // arguably shouldn't log anything on success
-          logger.info("No " + foundError + " found");
+          logger.info("No" + foundError + " found");
```
Comment (elharo) [inline enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java]: seems like there should be a space after No here.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ErrorType.java [hunk 8]
```diff
@@ -43,7 +43,7 @@
    * not public. If the source class is in the same package, the class or one of its enclosing types
    * is private.
    */
-  INACCESSIBLE_CLASS,
+  INACCESSIBLE_CLASS("is not accessible class"),
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ErrorType.java]: I don't think you need "class: at the the end, but if you do then "is not an accessible class"

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassFile.java (hunk 6 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassFile.java]: Nice. Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ErrorType.java (hunk 8 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ErrorType.java]: Removed "class" at the end.

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java (hunk 7 above)
Comment (suztomo) [reply enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java]: Fixed
(The space was in `foundError` but it was confusing.)

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java [hunk 9]
```diff
@@ -123,34 +128,32 @@ public void execute(@Nonnull EnforcerRuleHelper helper) throws EnforcerRuleExcep
...
+          symbolProblems = symbolProblems.entries().stream().filter(
+              entry -> classReferenceGraph.isReachable(entry.getKey().getClassName())).collect(
+              ImmutableSetMultimap.toImmutableSetMultimap(
+                  Entry::getKey,
+                  Entry::getValue
+              )
+          );
         }
+        // Count unique SymbolProblems
+        int errorCount = Sets.newHashSet(symbolProblems.values()).size();
```
[the remark this answers is not in the document]
Comment (suztomo) [reply enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java]: This is to be updated in next PR.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MethodSymbol.java [hunk 10]
```diff
@@ -81,11 +81,7 @@ public int hashCode() {
 
   @Override
   public String toString() {
-    return MoreObjects.toStringHelper(this)
-        .add("className", getClassName())
-        .add("name", name)
-        .add("descriptor", descriptor)
-        .add("isInterfaceMethod", isInterfaceMethod)
-        .toString();
+    return (isInterfaceMethod ? "Interface " : "") + getClassName() + "'s method " + Utility
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MethodSymbol.java]: Yes, reference to an interface method cannot be resolved by a method in a (non-interface) class with same name.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java [hunk 11]
```diff
@@ -68,10 +67,6 @@ public int hashCode() {
 
   @Override
   public String toString() {
-    return MoreObjects.toStringHelper(this)
-        .add("className", getClassName())
-        .add("name", name)
-        .add("descriptor", descriptor)
-        .toString();
+    return getClassName() + "'s field " + name;
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java]: Initially I thought about that but I kept this in the same form as MethodSymbol's toString where the output includes signature.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/MethodSymbolTest.java [hunk 12]
```diff
@@ -49,14 +49,26 @@ public void testNull() {
   public void testMethodSymbolEquality() {
     new EqualsTester()
         .addEqualityGroup(
-            new MethodSymbol("java.lang.Object", "equals", "foo", false),
-            new MethodSymbol("java.lang.Object", "equals", "foo", false))
-        .addEqualityGroup(new MethodSymbol("java.lang.Object", "hashCode", "foo", false))
-        .addEqualityGroup(new MethodSymbol("Object", "equals", "foo", false))
-        .addEqualityGroup(new MethodSymbol("java.lang.Object", "equals", "bar", false))
-        .addEqualityGroup(new MethodSymbol("java.lang.Object", "equals", "foo", true))
+            new MethodSymbol("java.lang.Object", "equals", "(I)Lcom.Foo;", false),
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/MethodSymbolTest.java]: Updated in #641

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/MethodSymbolTest.java [hunk 13]
```diff
@@ -49,14 +49,26 @@ public void testNull() {
...
+
+  @Test
+  public void testToString() {
+    MethodSymbol symbol = new MethodSymbol(
+        "java.lang.Object",
+        "equals",
+        "(Lcom/google/protobuf/Message;)Lio/grpc/MethodDescriptor$Marshaller;",
+        false);
+    assertEquals(
+        "java.lang.Object's method io.grpc.MethodDescriptor$Marshaller equals(com.google.protobuf.Message arg1)",
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/MethodSymbolTest.java]: Updated in #641

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (suztomo) [reply enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java]: It's been updated.

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java [hunk 14]
```diff
@@ -123,34 +128,32 @@ public void execute(@Nonnull EnforcerRuleHelper helper) throws EnforcerRuleExcep
...
 
-        String foundError = reportOnlyReachable ? "reachable error" : "error";
+        String foundError = reportOnlyReachable ? " reachable error" : " error";
         if (errorCount > 1) {
           foundError += "s";
         }
         if (errorCount > 0) {
           String message =
```
[the remark this answers is not in the document]
Comment (suztomo) [reply enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java]: It's been updated.

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java [hunk 15]
```diff
@@ -172,6 +175,28 @@ public void execute(@Nonnull EnforcerRuleHelper helper) throws EnforcerRuleExcep
     }
   }
 
+  @VisibleForTesting
+  static String formatSymbolProblems(ImmutableSetMultimap<ClassFile, SymbolProblem> symbolProblems) {
+    ImmutableSetMultimap<SymbolProblem, ClassFile> inversedSymbolProblems = symbolProblems
```
[the remark this answers is not in the document]
Comment (suztomo) [reply enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java]: The type of symbolProblems has been inversed. No need to create an inversed view any more.

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java [hunk 16]
```diff
@@ -172,6 +175,28 @@ public void execute(@Nonnull EnforcerRuleHelper helper) throws EnforcerRuleExcep
...
+        .inverse();
+    StringBuilder output = new StringBuilder();
+
+    for (SymbolProblem problem : inversedSymbolProblems.keySet()) {
```
[the remark this answers is not in the document]
Comment (suztomo) [reply enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java]: Updated in #641
